eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1327 @@
+      EknrBoxedSetInfo *set = g_ptr_array_index (sets, sets_index);
+
+      json_builder_begin_object (set_builder);
+
+      json_builder_set_member_name (set_builder, "id");
+      json_builder_add_string_value (set_builder, set->ekn_id);
+
+      json_builder_set_member_name (set_builder, "title");
+      json_builder_add_string_value (set_builder, set->title);
+
+      json_builder_set_member_name (set_builder, "featured");
+      json_builder_add_boolean_value (set_builder, set->featured);
+
+      json_builder_end_object (set_builder);
+
+      json_builder_add_value (sets_array_builder,
+                              json_builder_get_root (set_builder));
+    }
+
+  json_builder_end_array (sets_array_builder);
+
+  json_builder_set_member_name (metadata_builder, "sets");
+  json_builder_add_value (metadata_builder, json_builder_get_root (sets_array_builder));
+
+  json_builder_end_object (metadata_builder);
+
+  /* Stringify */
+  node = json_builder_get_root (metadata_builder);
+  generator = json_generator_new ();
+
+  json_generator_set_root (generator, node);
+  json_string = json_generator_to_data (generator, NULL);
+
+  return g_variant_new_string (json_string);
+}
+
+static char *
+get_wrapper_share_button_markup (const char  *network_name,
+                                 GError     **error)
+{
+  g_autofree char *svg_file_path = g_strdup_printf ("resource:///com/endlessm/knowledge/data/icons/scalable/apps/%s-symbolic.svg", network_name);
+  g_autoptr(GFile) svg_file = g_file_new_for_uri (svg_file_path);
+  g_autofree char *contents = NULL;
+
+  if (!g_file_load_contents (svg_file, NULL, &contents, NULL, NULL, error))
+    return NULL;
+
+  return g_strdup_printf ("<a class=\"share-action\"\n"
+                          "  onclick=\"window.webkit.messageHandlers.share_on_%s.postMessage(0)\">\n"
+                          "%s\n"
+                          "</a>\n",
+                          network_name,
+                          contents);
+}
+
+static GVariant *
+get_wrapper_share_actions_markup (const char *original_uri,
+                                  GError    **error)
+{
+  g_autofree char *facebook = NULL;
+  g_autofree char *twitter = NULL;
+  g_autofree char *whatsapp = NULL;
+  g_autofree char *markup = NULL;
+
+  if (original_uri == NULL || original_uri[0] == '\0')
+    return g_variant_new_string ("");
+
+  facebook = get_wrapper_share_button_markup ("facebook", error);
+
+  if (facebook == NULL)
+    return NULL;
+
+  twitter = get_wrapper_share_button_markup ("twitter", error);
+
+  if (twitter == NULL)
+    return NULL;
+
+  whatsapp = get_wrapper_share_button_markup ("whatsapp", error);
+
+  if (whatsapp == NULL)
+    return NULL;
+
+  markup = g_strdup_printf ("<div id=\"default-share-actions\" style=\"visibility: hidden;\">%s%s%s</div>",
+                            facebook,
+                            twitter,
+                            whatsapp);
+  return g_variant_new_string (markup);
```

Comment:
Use `g_variant_new_take_string (g_steal_pointer (&markup));` to avoid an extra copy.

Reply:
Handy, thanks!